Mathlib/Topology/Order/UpperSetTopology.lean

```diff
@@ -0,0 +1,283 @@
+
+variable {α}
+
+/-- If `α` is equipped with the upper set topology, then it is homeomorphic to
+`WithUpperSetTopology α`.
+-/
+def withUpperSetTopologyHomeomorph : WithUpperSetTopology α ≃ₜ α :=
+  WithUpperSetTopology.ofUpperSet.toHomeomorphOfInducing ⟨by erw [topology_eq α, induced_id]; rfl⟩
+
+lemma IsOpen_iff_IsUpperSet : IsOpen s ↔ IsUpperSet s := by
+  rw [topology_eq α]
+  rfl
+
+-- Alexandrov property, set formulation
+theorem IsOpen_sInter {S : Set (Set α)} (hf : ∀ s ∈ S, IsOpen s) : IsOpen (⋂₀ S) := by
+  rw [IsOpen_iff_IsUpperSet]
+  apply isUpperSet_sInter
+  intros s hs
+  rw [← IsOpen_iff_IsUpperSet]
+  exact hf _ hs
+
+-- Alexandrov property, index formulation
+theorem isOpen_iInter {f : ι → Set α} (hf : ∀ i, IsOpen (f i)) : IsOpen (⋂ i, f i) := by
+  rw [IsOpen_iff_IsUpperSet]
+  apply isUpperSet_iInter
+  intros i
+  rw [← IsOpen_iff_IsUpperSet]
+  exact hf i
+
+-- c.f. isClosed_iff_lower_and_subset_implies_LUB_mem
+lemma isClosed_iff_isLower {s : Set α} : IsClosed s ↔ (IsLowerSet s) := by
+  rw [← isOpen_compl_iff, IsOpen_iff_IsUpperSet,
+    isLowerSet_compl.symm, compl_compl]
+
+lemma isClosed_isLower {s : Set α} : IsClosed s → IsLowerSet s := fun h =>
+  (isClosed_iff_isLower.mp h)
+
+lemma closure_eq_lowerClosure {s : Set α} : closure s = lowerClosure s := by
+  rw [subset_antisymm_iff]
+  constructor
+  · apply closure_minimal subset_lowerClosure _
+    rw [isClosed_iff_isLower]
+    exact LowerSet.lower (lowerClosure s)
+  · apply lowerClosure_min subset_closure (isClosed_isLower isClosed_closure)
+
+/--
+The closure of a singleton `{a}` in the upper set topology is the right-closed left-infinite
+interval (-∞,a].
+-/
+@[simp] lemma closure_singleton {a : α} : closure {a} = Iic a := by
+  rw [closure_eq_lowerClosure, lowerClosure_singleton]
+  rfl
+
+end Preorder
+
+section maps
+
+variable [Preorder α] [Preorder β]
+
+lemma upperSetTopology_coinduced {t₁ : TopologicalSpace α} [UpperSetTopology α]
+    (hf : Monotone f) : coinduced f t₁ ≤ upperSetTopology' β := by
+  intro s hs
+  rw [isOpen_coinduced, IsOpen_iff_IsUpperSet]
+  exact (IsUpperSet.preimage hs hf)
+
+open Topology
+
+lemma Monotone_tfae {t₁ : TopologicalSpace α} [UpperSetTopology α]
+    {t₂ : TopologicalSpace β} [UpperSetTopology β] {f : α → β} :
+    List.TFAE
+    [ Monotone f,
+      Continuous f,
+      coinduced f t₁ ≤ t₂,
+      t₁ ≤ induced f t₂ ] := by
+  tfae_have 1 → 3
+  · intro hf s hs
+    rw [IsOpen_iff_IsUpperSet] at hs
+    exact upperSetTopology_coinduced hf _ hs
+  tfae_have 2 → 1
+  · intros hf a b hab
+    rw [← mem_Iic, ← closure_singleton, ← mem_preimage]
+    apply (Continuous.closure_preimage_subset hf {f b})
+    rw [← mem_Iic, ← closure_singleton] at hab
+    apply mem_of_mem_of_subset hab
+    apply closure_mono
+    rw [singleton_subset_iff, mem_preimage, mem_singleton_iff]
+  tfae_have 2 ↔ 4
+  · exact continuous_iff_le_induced
+  tfae_have 2 ↔ 3
+  · exact continuous_iff_coinduced_le
+  tfae_finish
```

Comment:
I would suggest doing the following:
```suggestion
protected lemma monotone_iff_continuous {t₁ : TopologicalSpace α} [UpperSetTopology α]
    {t₂ : TopologicalSpace β} [UpperSetTopology β] {f : α → β} :
    Monotone f ↔ Continuous f := by
  constructor
  · intro hf
    simp_rw [continuous_def, IsOpen_iff_IsUpperSet]
    exact fun _ hs ↦ IsUpperSet.preimage hs hf
  · intro hf a b hab
    rw [← mem_Iic, ← closure_singleton, ← mem_preimage]
    apply (Continuous.closure_preimage_subset hf {f b})
    rw [← mem_Iic, ← closure_singleton] at hab
    apply mem_of_mem_of_subset hab
    apply closure_mono
    rw [singleton_subset_iff, mem_preimage, mem_singleton_iff]
```
I don't think the `coinduced` and `induced` statements are worth putting in the TFAE because they are always one rewrite away anyways, and I feel like `rw [UpperSetTopology.monotone_iff_continuous, continuous_iff_coinduced_le]` is more readable than `rw [(UpperSetTopology.monotone_tfae).out 0 2]`.

Then you can also remove `upperSetTopology_coinduced` imo, first because it can easily be proven;
```lean
lemma upperSetTopology_coinduced {t₁ : TopologicalSpace α} [UpperSetTopology α]
    (hf : Monotone f) : coinduced f t₁ ≤ upperSetTopology' β := by
  rwa [← continuous_iff_coinduced_le, ← UpperSetTopology.monotone_iff_continuous]
```
and also because I think you can just work around it in #2508